### DISCUSSION TITLE
Mapping is OKed as Callable, but only if callable is defined as generic typevar?

### DISCUSSION THREAD

**Discussion by sizur** [microsoft/pyright#7727] [Q&A] [answered]
```python
type F[T: Callable[..., Any]] = T
m: F[Any] = cast(Mapping[Any, Any], ...)
```
Pyright is not grumpy about that.  Is this intentional (and if so, why), a bug, or a know limitation?

**Comment by erictraut** [comment]
The first line in your code defines a generic type alias that accepts a single type argument. The type annotation `F[Any]` replaces the type variable `T` with `Any`, so `F[Any]` is equivalent to `Any`. In other words, the second line is equivalent to:

```python
m: Any = cast(Mapping[Any, Any], ...)
```

Clearly, that should not result in any type violation because all types are compatible with `Any` in the Python type system.

**Reply by sizur** [reply]
And what about the type bound as specified in the generic type alias parameter definition as `Callable[..., Any]`?  Are you saying user should not expect the bound to be checked by the static type checker?

**Reply by erictraut** [reply]
The upper bound simply places a constraint on what types are allowed for `T`. Since you specified an upper bound of `Callable[..., Any]`, `T` must be a type that is compatible with `Callable[..., Any]`. That means, for example, that `T` cannot be `int`, so the type expression `F[int]` will result in a type error. However, `Any` is compatible with all types, so `Any` is an allowable value for `T`, and `F[Any]` does not result in a type error.

**Comment by sizur** [comment]
I see.  Does this break type safety, since F is defined as bound, but because bound doesn't propagate to `Any`, the variable's type breaks the type bound, right?  Is there a way to specify a type bound during assignment, without type information loss?  For example, for the following recursive type,
```python
type Struct[T] = int | str | Sequence[Struct[T]] | AbstractSet[Struct[T]] | Mapping[Struct[Any], Struct[T]]
```
how can one specify (A) a typevar bound to this type, and (B) a typevar bound to some narrowing of this type, since using `Any` as argument breaks the bound? (ignoring the fact that some type arguments there should be expected to be bound to `Hashable`)

**Reply by erictraut** [reply]
Sorry, but I don't understand your question. You're using terms like "propagate", "bound" and "assignment" in ways that are inconsistent with how we typically use these terms when talking about the Python type system, so I'm not sure what you mean. You may want to review the [typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html#type-variables-with-an-upper-bound) to familiarize yourself with the concept of type variables with upper bounds.

An upper bound on a type variable is simply a constraint. It indicates that the type argument associated with that type variable must be "consistent with" the upper bound. If you don't specify an upper bound, it is implicitly `object` which is the "top type" in the Python type system, so all types are consistent with it. The type `Any` is compatible with all types, so it can always be used as a type argument regardless of the upper bound. There's no way in the Python type system to specify that `Any` is disallowed as a type argument.

If you use `Any` in your code, type safety will be compromised. If you want your code to be completely type safe, don't use `Any`.

**Comment by sizur** [comment]
How can a generic type alias be used as a typevar bound, or it just can't and that's it?

**Reply by erictraut** [reply]
The upper bound expression cannot be parameterized with another type variable. That's explicitly disallowed in the [typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html#type-variables-with-an-upper-bound).

**Comment by sizur** [comment]
```python
type IntOrStr[T1: int | str] = T1
type Ext[T2: IntOrStr[Any] | float] = T2
```
Does the bound on Ext parameter effectively non-existent since you assert that `IntOrStr[Any]` is disregarding the `T1` bound and consequently equivalent to `Any`?

**Reply by erictraut** [reply]
The bound is a constraint. The type checker verifies that a supplied type argument is consistent with the upper bound. In your case, you're specifying a type argument of `Any`, and this satisfies the constraint because `Any` is consistent with `int | str`. No type error is reported. If you were to specify `IntOrStr[float]`, a type error would be reported because `float` is not consistent with `int | str`.

**Reply by sizur** [reply]
```python
def func[T: Ext[Any]](arg: T) -> T:
    assert isinstance(arg, (int, str, float))
    return arg
```
Can one rely on static type checker to ensure assertion doesn't fail or is T equivalent to `Any` according to current semantics?
If it's the second, what'd be a workaround, explicitly listing the union?

**Reply by sizur** [reply, marked as answer]
I get it.  Due `Any` semantics, I need to avoid parametric aliases for cases like this and use the non-parametric alias as constraint, like so:

```python
type IntOrStr = str | int
type Ext = IntOrStr | float
def func[T: Ext](arg: T) -> T:
    assert isinstance(arg, (int, str, float))
    return arg
```

Thank you for attempting to explain.

**Reply by erictraut** [reply]
`T` is not equivalent to `Any`. It is a type variable that has an upper bound of `Ext[Any]`, so it can take on any type that is compatible with `Ext[Any]`. Since `Ext[Any]` evaluates to `Any`, the value of `T` is effectively unconstrained and can take on any type. In other words, this upper bound is serving no purpose.

If your intent is to guarantee that `T` is consistent with `int`, `str`, or `float`, you should specify an upper bound of `str | float`. (The `int` is unnecessary in the union because `float` implies `int` in the Python type system.)

```python
def func[T: str | float](arg: T) -> T:
    ...

func(1) # OK
func(1.0) # OK
func("") # OK
func(3j) # Error
func([3]) # Error
```
